File: CSharp/SpaceVIL/SpaceVIL/Items/RadioButton.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Drawing;
using SpaceVIL.Core;
using SpaceVIL.Common;
using SpaceVIL.Decorations;

namespace SpaceVIL
{
    public class RadioButton : Prototype
    {
        internal class CustomIndicator : Indicator
        {
            internal override bool GetHoverVerification(float xpos, float ypos)
            {
                return false;
            }
        }

        static int count = 0;
        private TextLine _text_object;
        private CustomIndicator _indicator;

        /// <returns> RadioButton's indicator </returns>
        public Indicator GetIndicator()
        {
            return _indicator;
        }

        /// <summary>
        /// Constructs a RadioButton
        /// </summary>
        public RadioButton()
        {
            SetItemName("RadioButton_" + count);
            SetBackground(255, 255, 255, 20);
            SetSpacing(5, 0);
            EventKeyPress += OnKeyPress;
            count++;

            //text
            _text_object = new TextLine();
            // _text_object.IsFocusable = false;
            _text_object.SetItemName(GetItemName() + "_text_object");

            //indicator
            _indicator = new CustomIndicator();
            _indicator.IsFocusable = false;

            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.RadioButton)));
        }

        void OnKeyPress(object sender, KeyArgs args)
        {
            if (args.Key == KeyCode.Enter || args.Key == KeyCode.Space)
                EventMouseClick?.Invoke(this, new MouseArgs());
        }

        /// <summary>
        /// Set is mouse hover on the RadioButton
        /// </summary>
        public override void SetMouseHover(bool value)
        {
            base.SetMouseHover(value);
            _indicator.GetIndicatorMarker().SetMouseHover(IsMouseHover());
            UpdateState();
        }

        /// <summary>
        /// Initialization and adding of all elements in the RadioButton
        /// </summary>
        public override void InitElements()
        {
            //connect events
            _indicator.GetIndicatorMarker().EventToggle = null;
            EventMouseClick += (sender, args) =>
            {
                _indicator.GetIndicatorMarker().SetToggled(!_indicator.GetIndicatorMarker().IsToggled());
                if (_indicator.GetIndicatorMarker().IsToggled())
                    UncheckOthers(sender);
            };

            //adding
            AddItems(_indicator, _text_object);
        }

        /// <summary>
        /// Is RadioButton checked (boolean)
        /// </summary>
        public bool IsChecked()
        {
            return _indicator.GetIndicatorMarker().IsToggled();
        }
        public void SetChecked(bool value)
        {
            _indicator.GetIndicatorMarker().SetToggled(value);
        }

        private void UncheckOthers(object sender)
        {
            List<IBaseItem> items = GetParent().GetItems();
            foreach (var item in items)
            {
                if (item is RadioButton && !item.Equals(this))
                {
                    (item as RadioButton).GetIndicator().GetIndicatorMarker().SetToggled(false);
                }
            }
        }

        //text init
        /// <summary>
        /// Text alignment in the RadioButton
        /// </summary>
        public void SetTextAlignment(ItemAlignment alignment)
        {
            _text_object.SetTextAlignment(alignment);
        }

        /// <summary>
        /// Text margin in the RadioButton
        /// </summary>
        public void SetTextMargin(Indents margin)
        {
            _text_object.SetMargin(margin);
        }

        /// <summary>
        /// Text font parameters in the RadioButton
        /// </summary>
        public void SetFont(Font font)
        {
            _text_object.SetFont(font);
        }
        public void SetFontSize(int size)
        {
            _text_object.SetFontSize(size);
        }
        public void SetFontStyle(FontStyle style)
        {
            _text_object.SetFontStyle(style);
        }
        public void SetFontFamily(FontFamily font_family)
        {
            _text_object.SetFontFamily(font_family);
        }
        public Font GetFont()
        {
            return _text_object.GetFont();
        }

        /// <summary>
        /// Text in the RadioButton
        /// </summary>
        public void SetText(String text)
        {
            _text_object.SetItemText(text);
        }
        public String GetText()
        {
            return _text_object.GetText();
        }

        /// <summary>
        /// Text color in the RadioButton
        /// </summary>
        public void SetForeground(Color color)
        {
            _text_object.SetForeground(color);
        }
        public void SetForeground(int r, int g, int b)
        {
            _text_object.SetForeground(r, g, b);
        }
        public void SetForeground(int r, int g, int b, int a)
        {
            _text_object.SetForeground(r, g, b, a);
        }
        public void SetForeground(float r, float g, float b)
        {
            _text_object.SetForeground(r, g, b);
        }
        public void SetForeground(float r, float g, float b, float a)
        {
            _text_object.SetForeground(r, g, b, a);
        }
        public Color GetForeground()
        {
            return _text_object.GetForeground();
        }

        /// <summary>
        /// Set style of the RadioButton
        /// </summary>
        //style
        public override void SetStyle(Style style)
        {
            if (style == null)
                return;
            base.SetStyle(style);
            Style inner_style = style.GetInnerStyle("indicator");
            if (inner_style != null)
            {
                _indicator.SetStyle(inner_style);
            }
            inner_style = style.GetInnerStyle("textline");
            if (inner_style != null)
            {
                _text_object.SetStyle(inner_style);
            }
            SetForeground(style.Foreground);
            SetFont(style.Font);
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spacevil/RadioButton.java
package com.spacevil;

import com.spacevil.Core.*;

import java.awt.*;
import java.util.List;

public class RadioButton extends Prototype implements InterfaceHLayout {
    class CustomIndicator extends Indicator {
        @Override
        boolean getHoverVerification(float xpos, float ypos) {
            return false;
        }
    }

    private static int count = 0;
    private Label _text_object;
    private CustomIndicator _indicator;

    /**
     * @return RadioButton's indicator
     */
    public Indicator getIndicator() {
        return _indicator;
    }

    /**
     * Constructs a RadioButton
     */
    public RadioButton() {
        setItemName("RadioButton_" + count);
        count++;
        eventKeyPress.add(this::onKeyPress);

        // text
        _text_object = new Label();
        _text_object.isFocusable = false;
        _text_object.setItemName(getItemName() + "_text_object");

        // indicator
        _indicator = new CustomIndicator();
        _indicator.isFocusable = false;

        // setStyle(DefaultsService.getDefaultStyle("SpaceVIL.RadioButton"));
        setStyle(com.spacevil.Common.DefaultsService.getDefaultStyle(RadioButton.class));
    }

    /**
     * Constructs a RadioButton with text
     */
    public RadioButton(String text) {
        this();
        setText(text);
    }

    private void onKeyPress(InterfaceItem sender, KeyArgs args) {
        if (eventMouseClick != null && (args.key == com.spacevil.Flags.KeyCode.ENTER || args.key == com.spacevil.Flags.KeyCode.SPACE)) {
            eventMouseClick.execute(this, new MouseArgs());
        }
    }

    /**
     * Set is mouse hover on the RadioButton
     */
    @Override
    public void setMouseHover(boolean value) {
        super.setMouseHover(value);
        _indicator.getIndicatorMarker().setMouseHover(value);
        updateState();
    }

    // Layout rules
    /**
     * Add item to the RadioButton
     */
    @Override
    public void addItem(InterfaceBaseItem item) {
        super.addItem(item);
        updateLayout();
    }

    /**
     * Set width of the RadioButton
     */
    @Override
    public void setWidth(int width) {
        super.setWidth(width);
        updateLayout();
    }

    /**
     * Set X position of the RadioButton
     */
    @Override
    public void setX(int _x) {
        super.setX(_x);
        updateLayout();
    }

    /**
     * Update RadioButton's states (size and position)
     */
    public void updateLayout() {
        int offset = 0;
        int startX = getX() + getPadding().left;

        for (InterfaceBaseItem child : getItems()) {
            child.setX(startX + offset + child.getMargin().left);
            if (child.getWidthPolicy() == com.spacevil.Flags.SizePolicy.EXPAND) {
                child.setWidth(getWidth() - offset /*- child.getMargin().left - child.getMargin().right*/);
            }
            offset += child.getWidth() + getSpacing().horizontal;
        }
    }

    // text init
    /**
     * Text alignment in the RadioButton
     */
    public void setTextAlignment(com.spacevil.Flags.ItemAlignment... alignment) {
        _text_object.setTextAlignment(alignment);
    }
    public void setTextAlignment(List<com.spacevil.Flags.ItemAlignment> alignment) {
        _text_object.setTextAlignment(alignment);
    }

    /**
     * Text margin in the RadioButton
     */
    public void setTextMargin(com.spacevil.Decorations.Indents margin) {
        _text_object.setMargin(margin);
    }

    /**
     * Text font parameters in the RadioButton
     */
    public void setFont(Font font) {
        _text_object.setFont(font);
    }
    public void setFontSize(int size) {
        _text_object.setFontSize(size);
    }
    public void setFontStyle(int style) {
        _text_object.setFontStyle(style);
    }
    public void setFontFamily(String font_family) {
        _text_object.setFontFamily(font_family);
    }
    public Font getFont() {
        return _text_object.getFont();
    }

    /**
     * Text in the RadioButton
     */
    public void setText(String text) {
        _text_object.setText(text);
    }
    public String getText() {
        return _text_object.getText();
    }

    /**
     * Text color in the RadioButton
     */
    public void setForeground(Color color) {
        _text_object.setForeground(color);
    }
    public void setForeground(int r, int g, int b) {
        _text_object.setForeground(r, g, b);
    }
    public void setForeground(int r, int g, int b, int a) {
        _text_object.setForeground(r, g, b, a);
    }
    public void setForeground(float r, float g, float b) {
        _text_object.setForeground(r, g, b);
    }
    public void setForeground(float r, float g, float b, float a) {
        _text_object.setForeground(r, g, b, a);
    }
    public Color getForeground() {
        return _text_object.getForeground();
    }

    /**
     * Initialization and adding of all elements in the RadioButton
     */
    @Override
    public void initElements() {
        // events
        _indicator.getIndicatorMarker().eventToggle = null;
        InterfaceMouseMethodState btn_click = (sender, args) -> {
            _indicator.getIndicatorMarker().setToggled(!_indicator.getIndicatorMarker().isToggled());
            if (_indicator.getIndicatorMarker().isToggled())
                uncheckOthers(sender);
        };
        eventMouseClick.add(btn_click);

        // adding
        addItem(_indicator);
        addItem(_text_object);
    }

    private void uncheckOthers(InterfaceItem sender) {
        List<InterfaceBaseItem> items = getParent().getItems();
        for (InterfaceBaseItem item : items) {
            if (item instanceof RadioButton && !item.equals(this)) {
                ((RadioButton) item).getIndicator().getIndicatorMarker().setToggled(false);
            }
        }
    }

    /**
     * Set style of the RadioButton
     */
    // style
    @Override
    public void setStyle(com.spacevil.Decorations.Style style) {
        if (style == null)
            return;
        super.setStyle(style);

        setForeground(style.foreground);
        setFont(style.font);

        com.spacevil.Decorations.Style inner_style = style.getInnerStyle("indicator");
        if (inner_style != null) {
            _indicator.setStyle(inner_style);
        }
        inner_style = style.getInnerStyle("textline");
        if (inner_style != null) {
            _text_object.setStyle(inner_style);
        }
    }
}